### PR TITLE
Feature MIVOT (Model Instance in VOTable)

### LINKED ISSUE #12153: Accessing to the raw content of astropy.io.votable.tree.Resource(s)
### Description
In the frame of the VO data modeling efforts, we have been planning to annotate VOTable resources with data model meta-data.

These annotations are XML pieces matching a specific schema independent of the VOTable schema.  An annotated VOTable (work in progress) can be seen [here](https://github.com/ivoa-std/ModelInstanceInVot/blob/votable-schema/schema/validation/snippets/test_rich_instance_1_ok.xml) (branch `votable-schema`)

Annotations are not intended to be consumed by Astropy but rather by some affiliated  package. Anyway,  `astropy.io.votable` must be enabled to provide access to that XML for any end-client. 

In order to avoid to modify the current VOTable schema, these annotations will be enclosed in a Resource of type `meta` with a specific name space:

```xml
 <VOTable>
     <RESOURCE type=result>
         <RESOURCE type=meta>
             <dm-mapping:VODML ... My DM stuff ....>
        </RESOURCE>
       <TABLE/>       
       <TABLE/>
     </RESOURCE>
</VOTABLE>
```
This approach has been validated with regular XML parsers and validators and we need Astropy to support this new VOTable feature.
For this,  the`astropy.io.votable` API must be able to extract the content of such resources. This is apparently not possible yet.

The new endpoint should look like:
```python
votable = parse(votable_path)
   for resource in votable.resources:
       if resource.type == "results":
           for sub_resource in resource.resources:
               if sub_resource.type == "meta":
                   print("Here is my DM stuff")
                   annotation_content = sub_resource.content()
```

The resource content might be delivered as a DOM or/and as a string. The latest  would allow final clients to run their own XML processors.

### PR DESCRIPTION
**Pull request by somilia**
# Reading and Writing VO Model Annotations

## Introduction

Model Instances in VOTables ([MIVOT](https://ivoa.net/documents/MIVOT/20230620/REC-mivot-1.0.pdf)) defines a syntax for mapping VOTable data to any model serialized in VO-DML (Virtual Observatory Data Modeling Language). 

This annotation schema acts as a bridge between  data and  models. It associates both column/parameter metadata and VOTable data with data model elements (such as classes, attributes, types, etc.). It also complements VOTable data or metadata that may be missing from the table, e.g., coordinate system descriptions or curation tracing. 

The data model elements are organized in an independent annotation block complying with the MIVOT XML schema, which is added as an extra resource above the TABLE element. The MIVOT syntax allows a data structure to be described as a hierarchy of classes. It can also represent relationships and compositions between them. Furthermore, it can construct data model objects by aggregating instances from different tables of the VOTable.

## Astropy Implementation

The purpose of Astropy is not to process VO annotations. Instead, it allows related packages to get and set Mivot blocks from/into VOTables. Therefore, in this implementation, MIVOT annotations are both imported and exported as strings. The current implementation prevents client code from injecting into VOTables strings that are not Mivot serializations.

### `MivotBlock` Class:

- Mivot blocks are handled by the `MivotBlock` class in `tree.py.`
- A `MivotBlock` instance can only be carried by a resource with "type=meta."
- This instance holds the XML mapping block as a string.
- `MivotBlock` objects are instantiated by the Resource parser.
- The `MivotBlock` class has its logic that operates both parsing and I/O functionalities.

### Affected Modules:

- `tree.py`: Creation of the `MivotBlock`, addition of a `MivotBlock` instance to the `Resource` class as a property, and inclusion of a call to the MIVOT parser when the `Resource` parser encounters a VODML tag.
- `writer.py`: The function `string_element()` has been added to indent the `MivotBlock` XML before pushing it in the writer.
- `test_7_votable_pointer.py`: Some usage examples are described in `examples/mivotblock`.
- `exceptions.py`: Addition of warning (`ModelMappingSpecWarning`) and error (`E26`) classes related to MIVOT.

### Description

This pull request is for a new feature that allows the reading and the writing of model annotations in VOTable

Fixes #12153

- [ ] By checking this box, the PR author has requested that maintainers do **NOT** use the "Squash and Merge" button. Maintainers should respect this when possible; however, the final decision is at the discretion of the maintainer that merges the PR.

### CODE REVIEW

**Comment by somilia** [comment]
pre-commit.ci autofix

**Comment by lmichel** [comment]
> [pllim](https://github.com/pllim) added this to the [v6.0](https://github.com/astropy/astropy/milestone/98) milestone

I'm a new comer in the Astropy dev workflow and I do not get the meaning of this action.
Does it mean that the merge will occur with v6 and not before or that this PR will be regularly merged an then  reported on the v6 check points?

**Comment by pllim** [comment]
Re: https://github.com/astropy/astropy/pull/15390#issuecomment-1735932642

Not exactly. The milestone means this feature is currently slated for v6.0 release. We move the milestone if this PR does not get merged before feature freeze (also see https://github.com/astropy/astropy/wiki/Release-Calendar).

File: examples/mivotblock/test_7_votable_pointer.py
Comment (pllim) [inline examples/mivotblock/test_7_votable_pointer.py:1]: I have always wanted to kill off the example gallery. I see it as a failed experiment that had since been superseded by Learn Astropy efforts outside the core library. I'd rather we not add more stuff to it now. Also see https://github.com/astropy/astropy/issues/7242#issuecomment-1721460498

Can this example be integrated into `docs/io/votable` somewhere instead?

**Review by pllim — COMMENTED** [review]
Thanks! It is my understanding that PyVO is a stakeholder and its developers would review this in depth later. I am just leaving a general review.

Unless you are willing to let maintainer use "squash and merge" button for this PR, all the merge commits in your commit history here need to go (a rebase should do it in most cases).

Probably needs a What's New entry too.

If the API or behavior is expected to change after this code is released, please also add a warning box in the user documentation saying that the API or behavior is unstable and please be specific.

For the documentation that gets rendered to be read by users, please be consistent in capitalization (mivot vs MIVOT vs Mivot, VOtable vs VOTable, and so on). If different capitalization signifies different things (e.g., `astropy` vs Astropy), that is fine but you might have to explain.

When this PR is ready for final review, might want to add the "Extra CI" label too, just in case.

**Comment by pllim** [comment]
Also, please note that v6.0 feature freeze is planned for Oct 2023 in about a month (https://github.com/astropy/astropy/wiki/Release-Calendar). So your collaborators and stakeholders should start reviewing this sooner than later, or it might not make the cut.

**Comment by pllim** [comment]
p.s. For future reference, you could accept suggested changes in batch mode.

**Comment by somilia** [comment]
> Thanks! It is my understanding that PyVO is a stakeholder and its developers would review this in depth later. I am just leaving a general review.
> 
> Unless you are willing to let maintainer use "squash and merge" button for this PR, all the merge commits in your commit history here need to go (a rebase should do it in most cases).
> 
> Probably needs a What's New entry too.
> 
> If the API or behavior is expected to change after this code is released, please also add a warning box in the user documentation saying that the API or behavior is unstable and please be specific.
> 
> For the documentation that gets rendered to be read by users, please be consistent in capitalization (mivot vs MIVOT vs Mivot, VOtable vs VOTable, and so on). If different capitalization signifies different things (e.g., `astropy` vs Astropy), that is fine but you might have to explain.
> 
> When this PR is ready for final review, might want to add the "Extra CI" label too, just in case.

Thank you for the general review, changes will be coming soon.

Regarding the messy commit history, we would like a squash and merge if possible.

File: astropy/io/votable/tests/test_tree.py
Comment (somilia) [reply astropy/io/votable/tests/test_tree.py]: We think this function won't be useful beyond the test module.
We tried to do it in other way: using etree.XMLParser() or using re.sub() with regexep, in the end this function was more convenient and simpler.

File: astropy/io/votable/tests/test_tree.py
Comment (pllim) [inline astropy/io/votable/tests/test_tree.py:478]: I think you can just use our built-in `get_pkg_data_filename()` from `astropy.utils.data`.
```suggestion
    vpath = get_pkg_data_filename("data/test.order.xml")
```

**Comment by pllim** [comment]
You might not want to enable doctest on your What's New entry since it is trying to read a non-existent file. If you do want to use a test file you added, then you can do something like this instead:

```python
>>> from astropy.utils.data import get_pkg_data_filename

>>> filename = get_pkg_data_filename("data/votable.xml", package="astropy.io.votable.tests")  # If this file is in io/votable/tests/data, for example
# Then you can run the rest of the doctest with this filename
```

File: astropy/io/votable/tests/test_tree.py
Comment (somilia) [reply astropy/io/votable/tests/test_tree.py]: Thank you for the precision, changes have been made

**Comment by msdemlei** [comment]
Hm... while I agree that astropy-core probably shouldn't know too much about DMs as such, I'd say that the API implemented here is too poor to make much sense, for all I can see not even as a basis for later, more refined APIs: On reading, what do you expect me to do with your MivotBlock?  On writing, how do I come up with it?

For an alternative DM annotation, I once have implemented a somewhat more concrete API in https://github.com/msdemlei/astropy, where you could say things like

```
annotations = column.get_annotations("prefix:ClassName")
```

and then get back objects that would support attribute access for the various class members somewhat like this:

```
annotations[0].space.frame.orientation
```

I give you that because MIVOT is a lot more complex than the annotation scheme proposed in msdemlei/astropy, perhaps a less straightforward API may be called for – but then, having something like get_annotations even just for a subset of MIVOT would, I think, help people doing actual stuff once we have DM annotations out in the wild a lot more than the fairly opaque MivotBlock.

**Comment by lmichel** [comment]
> Hm... while I agree that astropy-core probably shouldn't know too much about DMs as such, I'd say that the API implemented here is too poor to make much sense, for all I can see not even as a basis for later, more refined APIs: On reading, what do you expect me to do with your MivotBlock? On writing, how do I come up with it?
> 
> For an alternative DM annotation, I once have implemented a somewhat more concrete API in https://github.com/msdemlei/astropy, where you could say things like
> 
> ```
> annotations = column.get_annotations("prefix:ClassName")
> ```
> 
> and then get back objects that would support attribute access for the various class members somewhat like this:
> 
> ```
> annotations[0].space.frame.orientation
> ```
> 
> I give you that because MIVOT is a lot more complex than the annotation scheme proposed in msdemlei/astropy, perhaps a less straightforward API may be called for – but then, having something like get_annotations even just for a subset of MIVOT would, I think, help people doing actual stuff once we have DM annotations out in the wild a lot more than the fairly opaque MivotBlock.

Markus,

You are right our API is very poor. Actually it does nothing except reading and writing XML strings and nothing in the ASTROPY code can work with that, this is true. These IO operations are implemented in AStropy just because the VOTable parser is here.
We chose this option last year after a few DM running meetings because we would like to avoid introducing any sort of model knowledge into Astropy. This is not the right place for this.
Actually the code you refer to is being implemented in PyVO.
Our roadmap is to open PyVO PRs just after this PR has been merged.

**Comment by pllim** [comment]
Anyone interested in this PR has any issue with changes to `tree.py` over at https://github.com/astropy/astropy/pull/15406 ?

**Comment by somilia** [comment]
> Anyone interested in this PR has any issue with changes to `tree.py` over at #15406 ?

After checking the lines, this PR shouldn't conflict or impact MIVOT PR.

File: astropy/io/votable/tree.py
Comment (tomdonaldson) [inline astropy/io/votable/tree.py:3627]: This seems to be returning an actual `MivotBlock` instance rather than a string as indicated in the function doc string.  Although the `MivotBlock` can be treated as a string in many ways, should the doc string be adjusted to say that the return is a `MivotBlock`?

File: astropy/io/votable/tests/test_tree.py
Comment (tomdonaldson) [inline astropy/io/votable/tests/test_tree.py:209]: The filename `mivot_annnotated_table.xml` has `nnn` instead of `nn` which is probably a typo.  If so, to avoid confusion, both the filename and this line should be fixed.

**Review by tomdonaldson — CHANGES_REQUESTED** [review]
Thanks @lmichel and @somilia!  This MIVOT support seems very complete to me, and the tests and documentation look good.  Other than a couple minor comments from me above, I think this is ready to go.

**Comment by somilia** [comment]
> Thanks @lmichel and @somilia! This MIVOT support seems very complete to me, and the tests and documentation look good. Other than a couple minor comments from me above, I think this is ready to go.

Thank you for your review and comments, changes have been made.

**Review by tomdonaldson — APPROVED** [review]

**Comment by pllim** [comment]
p.s. Once VO reviewers approved, I would still go through the code one more time to make sure it adheres to Astropy standards. So please do not merge yet. But let me know if I can merge when I am happy with it. Thanks!

File: astropy/io/votable/tree.py
Comment (pllim) [inline astropy/io/votable/tree.py:3476]:
```suggestion
        Regular parser similar to others VOTable components.
```

File: astropy/io/votable/tree.py
Comment (pllim) [inline astropy/io/votable/tree.py:3461]:
```suggestion
    def content(self):
```

File: astropy/io/votable/tree.py
Comment (pllim) [inline astropy/io/votable/tree.py:3513]:
```suggestion
        if not self._content.startswith("<"):
```

File: astropy/io/votable/tree.py
Comment (pllim) [inline astropy/io/votable/tree.py:3621]:
```suggestion
        Otherwise, it is taken from the host resource.
```

File: docs/changes/io.votable/15390.feature.rst
Comment (pllim) [inline docs/changes/io.votable/15390.feature.rst:1]:
```suggestion
Added MIVOT feature through the ``MivotBlock`` class
```

File: docs/io/votable/index.rst
Comment (pllim) [inline docs/io/votable/index.rst:495]:
```suggestion
- A MivotBlock instance can only be carried by a resource with "type=meta".
```

File: docs/io/votable/index.rst
Comment (pllim) [inline docs/io/votable/index.rst:468]: Ref label for intersphinx:

```suggestion
.. _votable_mivot:

Reading and writing VO model annotations
```

File: docs/whatsnew/6.0.rst
Comment (pllim) [inline docs/whatsnew/6.0.rst:1]: The contents here just basically repeats the stuff you added in `docs/io/votable/index.rst`. Maybe make this more concise and point to your other doc using intersphinx like so?

```
:ref:`votable_mivot`
```

**Review by pllim — COMMENTED** [review]
Just need a little more minor clean-ups, otherwise LGTM. Thanks!

File: docs/whatsnew/6.0.rst
Comment (somilia) [reply docs/whatsnew/6.0.rst:1]: You're right, I summarized the what's new, and add an intersphinx as you recommended.
Is it normal that one of the test in the CI is failing ?

**Comment by pllim** [comment]
Something went wrong with the commit history. I thought @bsipocz squashed the stuff into 14 commits last Friday. So with 2 new commits, it should have been 16. But I see 56 commits now. And the CI failure does not make sense. So something is wrong.

**Comment by bsipocz** [comment]
I suspect the old branch got force-pushed back rather than the rebased version pulled and updated. I still have the 14-commit version around locally and can push it back. As I recall that one was passing CI and was waiting for your review., but was otherwise ready to go.

**Comment by bsipocz** [comment]
Ahh, looking at the branch structure, the second to last commit is still correct, the problem is with the last one only, as I suspect that was done locally on the outdated branch while second to last was done directly on github.

**Comment by pllim** [comment]
devdeps failure is unrelated (numpy-dev again), so merging. Thanks for your patience and contribution, all!

**Comment by lmichel** [comment]
Great, tanks to all of the reviewers.

**Pull request merged**
